src/main/java/org/datadog/jenkins/plugins/datadog/util/git/GitUtils.java

```diff
@@ -111,6 +115,39 @@ public static GitMetadata buildGitMetadata(@Nullable final GitClient gitClient)
         }
     }
 
+    /**
+     * When checking out a repository for a multi-branch pipeline,
+     * Jenkins does so in "detached HEAD" state.
+     * This means that a specific commit is checked out, rather than a branch.
+     * To work around this the plugin examines available branches
+     * to see which one of them points to the checked out commit.
+     * <p>
+     * It is possible that multiple branches point to the same commit.
+     * In which case the logic in this method tries to use a "hint" extracted from environment variables.
+     */
+    private static String chooseBranch(@Nonnull Collection<String> branches, @Nullable String branchHint, String buildName) {
+        if (branchHint != null && branches.contains(branchHint)) {
+            // multiple branches point to the checked out commit,
+            // and one of them is the same as the branch that is set in the env vars
+            return branchHint;
+
+        } else if (branches.size() > 1) {
+            // multiple branches point to the checked out commit,
+            // and none of them is the same as the branch that is set in the env vars,
+            // we are taking a guess at this point
+            LOGGER.warning("Build " + buildName + " has multiple Git branches matched: " + branches);
+            return branches.iterator().next();
+
+        } else if (branches.size() == 1) {
+            // only one branch points to the checked out commit
+            // (also possible that the repo was checked out properly, and not in a detached HEAD state)
+            return branches.iterator().next();
```

Comment:
If I understood correctly, here we're always returning the first branch (`iterator().next()`) for both cases  `branches.size() >= 1`. If that's the case why we need to store all branches?

Also, could we simplify this block using `>=1`? And, should we use `List<>` instead of `Collection<>` to use `get(0)`?

Reply:
All branches were needed in order to check if the list contains the branch that is set in the env vars. I removed this method in favour of resolving the branch inside the repo callback, but the logic remained essentially the same.